### PR TITLE
Sort hour of code tutorials by popularity

### PR DESCRIPTION
**Pull request by molly-moen**
# Description
Update hourofcode.com/learn, code.org/learn, and csedweek.org/learn to sort tutorials by popularity when in the "post-hoc" mode. Current mode for production is "post-hoc".
Screen Shot 2020-01-08 at 10 57 30 AM

## Links

- [jira](https://codedotorg.atlassian.net/browse/PLC-684)

## Testing story
Validated all pages sort as expected during "post-hoc" mode. During "pre-hoc" mode only code.org/learn will sort by popularity as was the behavior previously.

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: pegasus/sites.v3/code.org/public/learn/index.haml [hunk 1]
```diff
@@ -120,8 +120,8 @@ style_min: false
   - tutorial[:string_detail_platforms]              = hoc_s(prefix + "string_detail_platforms")
   - tutorial[:string_detail_programming_languages]  = hoc_s(prefix + "string_detail_programming_languages")
 
-- sort_by_popularity = request.site == 'code.org'
-- sort_by_popularity &&= [false, "pre-hoc"].include?(DCDO.get("hoc_mode", CDO.default_hoc_mode))
+- hoc_mode = DCDO.get("hoc_mode", CDO.default_hoc_mode)
+- sort_by_popularity = [false, "post-hoc"].include?(hoc_mode) || (request.site == 'code.org' && hoc_mode == 'pre-hoc')
```
Comment (islemaster) [inline pegasus/sites.v3/code.org/public/learn/index.haml]: This changes the behavior when `request.site` is `csedweek.org` and `hoc_mode` is `false`. Before, `sort_by_popularity` would be `false` in that scenario; now it's `true`.  I'm pretty sure the fix is:

```suggestion
- sort_by_popularity = ("post-hoc" == hoc_mode) || (request.site == 'code.org' && [false, 'pre-hoc'].include? hoc_mode)
```

**Review by islemaster — CHANGES_REQUESTED** [review]
This logic for the sort order is starting to get a little complex!  And it's duplicated between the `code.org` and `hourofcode.com` helpers.  Let's extract it someplace and write tests for it.

I think this is a good candidate for a class method on the `Tutorials` class in [database.rb](https://github.com/code-dot-org/code-dot-org/blob/8ef326cb38a8233b569f8204d18f3f36e49db83c/pegasus/src/database.rb#L9).  There are [already some tests for that class](https://github.com/code-dot-org/code-dot-org/blob/df9d9c92acbba6b4dde4e414d0bb0eb24595692f/pegasus/test/test_tutorials.rb#L5).  Maybe a signature that looks like this?

```rb
def self.sort_by_popularity?(site, hoc_mode)
```

Let me know if you'd like to chat about this in person.

File: pegasus/sites.v3/code.org/public/learn/index.haml (hunk 1 above)
Comment (molly-moen) [reply pegasus/sites.v3/code.org/public/learn/index.haml]: makes sense, incorporated that change into new method

File: pegasus/test/test_tutorials.rb [hunk 2]
```diff
@@ -26,4 +26,34 @@ def test_tutorials_contents_immutable
     tutorial[:image] = 'xyz'
     refute_equal Tutorials.new(:tutorials).contents('code.org').first[:image], tutorial[:image]
   end
+
+  def test_sort_by_popularity_code_org
+    code_org = "code.org"
+
+    assert Tutorials.sort_by_popularity?(code_org, "post-hoc")
+    assert Tutorials.sort_by_popularity?(code_org, "pre-hoc")
+    assert Tutorials.sort_by_popularity?(code_org, false)
```
Comment (islemaster) [inline pegasus/test/test_tutorials.rb:35]: Could also add `soon-hoc` and `actual-hoc` to cover the whole cycle of HOC-modes.  ([See this doc](https://docs.google.com/document/d/1npRdcuOhOPGghHXkBZain_7ZvRrqvizAbwT1gwlAaK8/edit))

**Review by islemaster — APPROVED** [review]
:tada: Looks great!

**Review by breville — COMMENTED** [review]
Thanks for jumping into this code!

**Pull request merged**